gnosis/cowswap
# Add dynamic announcements

# Summary

Part of #1192 

This PR allows to add announcements dynamically: 
![image](https://user-images.githubusercontent.com/2352112/128548514-d4b244b8-a79e-4a03-bdfa-d781a3565d1e.png)

The way it works, is that it will load the content for the announcement from an external source. 
* If there's no announcement (empty text or error loading), it won't show the banner.
* If there's some content, it will show them as shown in the image above

This source, for now is a file in one specific branch of the project.

## How to change the announcements
Modify one specific file in the `anouncements` branch: * https://github.com/gnosis/cowswap/blob/announcements/docs/announcements.md



## What happens when the user close the announcement
The announcement can be closed by the user.

Once the user closes one specific announcement, it will never be shown again. 

The way this works technically is: The content of the annoncement will be hashed using a very basic hashing function (https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript).
The app will persist the information that the user closed that specific announcement so it's not displayed again.

New announcements will have a different hash, for this reason they will be shown even if the user has closed other messages.

## Implementation details
The banner is implemented reusing `URLWarning` (for simplicity).

The way it works is, if there's one announcement, it will be shown. If not, the component has the default behaviour.

Also, regarding the state. Uniswap had already a `user` state where they would save these kind of internal state for the user preferences. I don't want to hack those parts directly, so I created a new state `userMod` for that. 


# To Test

1. Load app
2. See that the announcement is displayed
3. Close announcement
4. Notice that now, if you update the app is not shown anymore
5. Update the announcements
6. Observe that new messages are displayed

## Caveats
* We shouldn't reuse messages if we want all users to see them. This is because it uses a hashing function over its content to identify it and keep track of the closed ones. I think it's good enough for now. In the future this can use some other service to back this info. 
* Another thing to take into account is that github RAW files have a cache of 5min, so even if you change the content, it won't be updated for some time.
* Lastly, there's no validation in the length of the announcement. It can break the app if it's too long. Limit yourselves to small concise messages.
